### PR TITLE
improved HUD (barcharts and joblist) failures page with branch information

### PR DESCRIPTION
**Pull request by daizeng1984**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #969

Pytorch CI's [failure page](https://hud.pytorch.org/failure/test_aot_autograd) can show us failures dashboard for information. However, that page shows errors across all branches, spanning both master and private branches, and doesn't offer any hint of which branch a given failure occurred on. Knowing if failures are happening on the master branch or on private branches would help debug issues much faster.

This change on the failures page, will highlight failures that occurred on the master branch in barcharts so that we could easily see master branch's problem but also show the branch for the failure job detailed list.

This new view could help better identify master branch issues and also get a feeling (distribution) of overall job failures.

# Old view:
image

image

# New view:
image
image

### CODE REVIEW

File: torchci/pages/failure/[capture].tsx [hunk 1]
```diff
@@ -23,38 +24,145 @@ function FailureInfo({
...
     const time = dayjs().local().subtract(i, "day").format("MM/D");
-    dayBuckets.set(time, 0);
+    dayBuckets.set(time, { highlighted: 0, other: 0, total: 0 });
+    branchHistogramByDay.set(time, new Map());
   }
 
-  samples.forEach((job) => {
+  // we highlight master branch
+  const highlighted = new Set<string>();
+  highlighted.add("master");
```
Comment (ZainRizvi) [inline torchci/pages/failure/[capture].tsx:40]: Nice!

File: torchci/components/JobSummary.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ export default function JobSummary({ job }: { job: JobData }) {
     <>
       <JobConclusion conclusion={job.conclusion} />
       <a href={job.htmlUrl}> {job.jobName} </a>
+      <b> [{job.branch}] </b>
```
Comment (ZainRizvi) [inline torchci/components/JobSummary.tsx:9]: Can we only bold the branch name if it's the "master" branch?

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
This looks fantastic! I especially like what you've done with the bar chart, that's going above and beyond :D 

Just one small tweak to only bold the master branches, other than that this looks great!

File: torchci/components/JobSummary.tsx [hunk 3]
```diff
@@ -6,6 +6,7 @@ export default function JobSummary({ job }: { job: JobData }) {
     <>
       <JobConclusion conclusion={job.conclusion} />
       <a href={job.htmlUrl}> {job.jobName} </a>
+      <b> {job.branch} </b>
```
Comment (ZainRizvi) [inline torchci/components/JobSummary.tsx:9]: > Can we only bold the branch name if it's the "master" branch?

I think there was some misunderstanding with my [original comment](https://github.com/pytorch/test-infra/pull/969#discussion_r1008241570).

I meant that when `job.branchname == "master"`, the html here should look like:
`<b> [{job.branch}] </b>` (how it originally looked)

While at all other times it should effectively look like:
`<span> [{job.branch}] </span>` (or something like that

**Review by ZainRizvi — CHANGES_REQUESTED** [review]

File: torchci/components/JobSummary.tsx [hunk 4]
```diff
@@ -1,11 +1,16 @@
 import { JobData } from "lib/types";
 import JobConclusion from "./JobConclusion";
 
-export default function JobSummary({ job }: { job: JobData }) {
+export default function JobSummary({ job, highlight }: { job: JobData; highlight: boolean; }) {
   return (
     <>
       <JobConclusion conclusion={job.conclusion} />
       <a href={job.htmlUrl}> {job.jobName} </a>
+      {highlight ? <b> [{job.branch}] </b> : <span>[{job.branch}]</span> }
```
Comment (ZainRizvi) [inline torchci/components/JobSummary.tsx:9]: One more bug: This component is shared with other pages, and those pages aren't setup to receive the branch (and they don't need to be).

However, the branch being null there is resulting in those pages having empty brackets:

Example: https://torchci-git-gh-daizeng19841head-fbopensource.vercel.app/pytorch/pytorch/commit/1da5aeb97b73664ff0fe2f4bb48379655cede969

image

**Review by ZainRizvi — CHANGES_REQUESTED** [review]

File: torchci/components/JobSummary.tsx (hunk 4 above)
Comment (daizeng1984) [reply torchci/components/JobSummary.tsx]: Thanks for catching that! I've added the null/empty check for job.branch

File: torchci/components/JobSummary.tsx [hunk 5]
```diff
@@ -1,11 +1,16 @@
...
-export default function JobSummary({ job }: { job: JobData }) {
+export default function JobSummary({ job, highlight }: { job: JobData; highlight: boolean; }) {
   return (
     <>
       <JobConclusion conclusion={job.conclusion} />
       <a href={job.htmlUrl}> {job.jobName} </a>
+      {!job.branch? <></> : (highlight ? <b> [{job.branch}] </b> : <span>[{job.branch}]</span>) }
     </>
   );
 }
```
Comment (ZainRizvi) [inline torchci/components/JobSummary.tsx:12]: All these unary operators are making it hard to read :) 
What about replacing this with something like the following to help with readability:

Some more optimization may be possible to selectively choose the tag type.

```
  var branchName = <></>
  if (job.branch) {
    if (highlight) {
      branchName = (<b> [{job.branch}] </b>)
   } else {
      branchName = (<span>[{job.branch}]</span>)
   }
  }
  return (
    <>
      <JobConclusion conclusion={job.conclusion} />
      <a href={job.htmlUrl}> {job.jobName} </a>
      {branchName}
    </>
  );
}
```

**Review by ZainRizvi — CHANGES_REQUESTED** [review]

**Review by ZainRizvi — APPROVED** [review]
Thanks for making this change!

**Pull request merged**